### PR TITLE
fix: token detection and import

### PR DESCRIPTION
**Pull request by bergeron**
## **Description**

Fixes issues with detecting and manually importing tokens, caused by recent controller changes.

Hotfix for 11.14.0

This state publish had an undefined variable, and had the wrong function signature:

Screenshot 2024-03-28 at 6 04 41 PM

Which prevented the controllers from processing account switches.

A test was failing because RPC requests were sent before onboarding.  Delayed the enabling of token detection controller to `triggerNetworkrequests`.

Also patching this fix for detection during account switch on mainnet when autodetection=off:
https://github.com/MetaMask/core/pull/4133

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23798?quickstart=1)

## **Related issues**

Fixes:
- https://github.com/MetaMask/metamask-extension/issues/23526
- https://github.com/MetaMask/metamask-extension/issues/23612

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by salimtb — DISMISSED** [review]

**Comment by salimtb** [comment]
@bergeron can you check why your e2e tests are failing ?

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -2507,8 +2507,8 @@ export default class MetamaskController extends EventEmitter {
 
       // TODO: Remove once the preferences controller has been replaced with the core monorepo implementation
       this.controllerMessenger.publish(
-        `${this.preferencesController.name}:stateChange`,
-        [state, []],
+        'PreferencesController:stateChange',
+        state,
```
Comment (bergeron) [inline app/scripts/metamask-controller.js:2511]: The old code's state publish was coming through as an array, which did not destruct correctly.  I don't know if the proper fix is to publish an object, or deconstruct the first element of the array instead.

image

Comment (MajorLift) [inline app/scripts/metamask-controller.js:2511]: These would be the arguments expected by this `publish` call:

```suggestion
        state,
        [],
```

or alternatively,

```suggestion
        ...[state, []],
```

For reference, this is the signature of the `publish` method of `RestrictedControllerMessenger`:

```ts
publish<EventType extends Event['type'] & NamespacedName<Namespace>>(
    event: EventType,
    ...payload: ExtractEventPayload<Event, EventType>
  )
```

**Comment by MajorLift** [comment]
Probably unrelated to the bugs, but this is another error related to the preferences-controller that was introduced in #22928.

```diff
diff --git a/app/scripts/metamask-controller.js b/app/scripts/metamask-controller.js
index 89905caaef..ddecaf42f5 100644
--- a/app/scripts/metamask-controller.js
+++ b/app/scripts/metamask-controller.js
@@ -500,8 +500,8 @@ export default class MetamaskController extends EventEmitter {
 
     const preferencesMessenger = this.controllerMessenger.getRestricted({
       name: 'PreferencesController',
-      allowedActions: ['PreferencesController:getState'],
-      allowedEvents: ['PreferencesController:stateChange'],
+      allowedActions: [],
+      allowedEvents: [],
     });
 
     this.preferencesController = new PreferencesController({
```

**Comment by bergeron** [comment]
Test failing because we're sending RPC requests before onboarding finishes.  Test was recently added while detection was broken.  Will see if I can delay the controller's start.

**Review by MajorLift — APPROVED** [review]
LGTM! Thanks for making these fixes.

**Review by sahar-fehri — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]

**Pull request merged**